File: server/src/models/ExecutionHistory.ts
import mongoose, { Schema, Document, Model } from 'mongoose'

// ✅ TypeScript interface for ExecutionHistory documents
export interface IExecutionHistory extends Document {
  user: mongoose.Types.ObjectId | null
  roomId?: string | null
  language: string
  code: string
  stdin?: string
  output?: string
  stderr?: string
  time?: number | null
  createdAt: Date
  updatedAt: Date
}

// ✅ Define schema
const executionHistorySchema = new Schema<IExecutionHistory>(
  {
    // Linked user (who executed the code)
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      default: null,
    },

    // Optional room (for collaborative sessions)
    roomId: {
      type: String,
      default: null,
    },

    // Language used (restricted for data consistency)
    language: {
      type: String,
      required: true,
      enum: ['javascript', 'typescript', 'python', 'cpp', 'java', 'c'],
    },

    // The code that was executed
    code: {
      type: String,
      required: true,
    },

    // Optional standard input
    stdin: {
      type: String,
      default: '',
    },

    // Execution output
    output: {
      type: String,
      default: '',
    },

    // Error output, if any
    stderr: {
      type: String,
      default: '',
    },

    // Execution time in milliseconds (optional)
    time: {
      type: Number,
      default: null,
    },
  },
  { timestamps: true }
)

// ✅ Indexes for fast queries (by user or room)
executionHistorySchema.index({ user: 1, createdAt: -1 })
executionHistorySchema.index({ roomId: 1, createdAt: -1 })

// ✅ Create and export the model
export const ExecutionHistory: Model<IExecutionHistory> =
  mongoose.model<IExecutionHistory>('ExecutionHistory', executionHistorySchema)

File: server-spring/src/main/java/com/codeeditor/server/model/ExecutionHistory.java
package com.codeeditor.server.model;

import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.CompoundIndex;
import org.springframework.data.mongodb.core.index.CompoundIndexes;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.Instant;
import org.springframework.data.mongodb.core.mapping.Field;
import org.springframework.data.mongodb.core.mapping.FieldType;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "executionhistories")
@CompoundIndexes({
    @CompoundIndex(name = "user_created_idx", def = "{'user': 1, 'createdAt': -1}"),
    @CompoundIndex(name = "roomId_created_idx", def = "{'roomId': 1, 'createdAt': -1}")
})
public class ExecutionHistory {

    @Id
    private String id;

    @Field(targetType = FieldType.OBJECT_ID)
    private String user; // References User._id (nullable)

    private String roomId; // Optional room ID

    private String language; // javascript, typescript, python, cpp, java, c

    private String code;

    private String stdin = "";

    private String output = "";

    private String stderr = "";

    private Double time; // Execution time in milliseconds

    @CreatedDate
    private Instant createdAt;

    @LastModifiedDate
    private Instant updatedAt;
}
